### ISSUE TITLE
Squirrel installers race for SquirrelTemp resulting in bad installations

### ISSUE THREAD

**Issue by riverar** [Squirrel/Squirrel.Windows#1853]
**Squirrel version(s)**
squirrel.windows.2.0.1 and possibly others

**Description**
Squirrel uses `%localappdata%\SquirrelTemp` for various servicing tasks. Squirrel seems to allow simultanous installations to occur yet does not silo the temporary folder on a per-app basis. This appears to result in races for files in the temporary folder and can result in incorrect installations of various applications, wrong shortcuts, incorrect schema launches, etc. These installations are very difficult to repair and detremential to the user experience.

I believe this meets the bug bar for resolution under the _Focus on the User Experience_ principle (https://github.com/Squirrel/Squirrel.Windows/issues/1470#issuecomment-486410297) and the product's tag line _Squirrel: It's like ClickOnce but Works™_.

**Steps to recreate**
I highly recommend the use of a virtual machine or Windows Sandbox.

1. Download Teams_windows_x64.exe from https://statics.teams.microsoft.com/production-windows-x64/1.6.00.6754/Teams_windows_x64.exe
2. Download Scenario.zip (attached here)
[scenario.zip](https://github.com/Squirrel/Squirrel.Windows/files/11025009/scenario.zip)
3. Extract Scenario.zip to a temporary location and move Teams_windows_x64.exe into this location so all three files are side-by-side
4. Execute `scenario.cmd` and wait until completion
5. Run `ms-settings:installed-apps` or navigate to Settings > Apps > Installed Apps
6. Observe the included demo app with publisher Rafael is associated with Microsoft Teams
7. Open Start and type `myapp` to begin a search.
8. Observe the included demo app has incorrect shortcuts

> Tip: Alternatively, you can shorten the timeout and remove the process kill to demonstrate a real race. But results may vary across machine speeds.

**Screenshots**
![image](https://user-images.githubusercontent.com/475132/226517785-c7b9c221-0cf9-46ce-b478-4368a9cbd0ba.png)
![image](https://user-images.githubusercontent.com/475132/226517916-4229e5aa-de04-4eee-819b-d722b1a88221.png)

**Expected behavior**
It should not be possible for Squirrel apps to read from or write over each other

**Actual behavior**
Squirrel apps compete for the temporary folder and behave indeterminately

**Proposal**
Use a product-specific (or randomly generated) path to silo apps from each other.

**Comment by anaisbetts** [comment]
While this is still a bug, MS Teams uses a forked, private copy of Squirrel that has several known bugs (some of which were self-inflicted) and we can do nothing to fix them

**Comment by riverar** [comment]
Interesting. Would the proposed silos at least shield the community from Teams and other apps using outdated Squirrel code? Or is this a set of unrelated issues?

**Comment by anaisbetts** [comment]
A quick review of [the code that I wrote ~9 years ago](https://github.com/Squirrel/Squirrel.Windows/blob/76c87af6f389bea69576c738ef46574b2945b0ba/src/Setup/UpdateRunner.cpp#L193) seems to indicate that this is not super easy because of how early in the process we are, though the "randomly generated" strategy might work. I'm.....not sure why I didn't do that in the first place actually! Probably because I was trying to make the C++ section of this code As Small As Possible

The other places we use SquirrelTemp during the update process, we actually already use the "generate a unique empty subdirectory" strategy

**Comment by anaisbetts** [comment]
@robmen ^^ a Good Idea, this should fix some of the issues that people have around machine-wide installers stepping on each other

**Comment by robmen** [comment]
Yeah. I'm finally getting to the end of my WiX v4 saga and should _finally_ have some mental energy for Squirrel.

**Comment by gwenchailleu** [comment]
Hello,
With several colleagues from my company, we chose squirrel to replace clikonce for the deployment of our products.
While coding a new application I came across #1853.
Do you think our help in resolving it would be useful to you? We could make correction proposals in the coming months...
Or do you prefer to keep your hands on this subject?

**Comment by riverar** [comment]
@gwenchailleu I would submit a PR with the solution documented above so we can get the ball rolling here. This is a critical bug and am quite surprised this still hasn't been resolved.

**Comment by gwenchailleu** [comment]
Hello, this is good news ! 
We coded our own correction of this issue but we had no time to share it. 
Are you interested in a PR ?
Regards

**Comment by dopry** [comment]
I am interested in a PR. I believe our users are encountering this when double launching the installer on accident. One installer will delete the temp folder out from other the other as corrupt. This is the error I consistently get when double launching the installer... 

![image](https://github.com/user-attachments/assets/10a03985-08a8-4afa-bf0e-1a6bfa387e76)

Then Open Setup Log opens a path that doesn't exist. 

but when I look in the SquirrelTemp folder I have Squirrel-Install.log and Squirrel-Install.1.log, that seem to indicate they're both trying to manipulate the contents of the SquirrelTemp and the AppData\{appName} folder. 


[Squirrel-Install.zip](https://github.com/user-attachments/files/18154271/Squirrel-Install.zip)

Both installers fail when launched in rapid succession like this.
